### ISSUE TITLE
Add autoregistration functionality to agents

### ISSUE THREAD

**Issue by ViktorOrda** [crowdsecurity/crowdsec#3319]
### What would you like to be added?

In this https://github.com/crowdsecurity/crowdsec/pull/3202 PR possibility to automatically register machines in LAPI was introduced. As far as I can see, there are no such **automatical** fuctionality for agents: as pointed in the [docs](https://docs.crowdsec.net/u/user_guides/machines_mgmt/#machine-auto-validation), after LAPI configuration one should manually register agent machine. We wanna use containerized solution with autoscaling, so agents autoregistration is crucial for us. 

I've tried several ways to implement such a solution locally:
1. Start agent container with LAPI disabled (and try to configure remote LAPI server in runtime) - agent container didn't start because of the next errors:
```
level=debug msg="running agent after 625.768181ms ms"
level=debug msg="[URL] POST http://0.0.0.0:8080/v1/watchers/login"
level=debug msg="resetting jwt token"
level=fatal msg="unable to start crowdsec routines: authenticate watcher (): Post \"http://0.0.0.0:8080/v1/watchers/login\": performing jwt auth: dial tcp 0.0.0.0:8080: connect: connection refused" 
```

2. Start agent container with LAPI enabled, register it to the remote LAPI and restart agent (as described in the [blog post](https://www.crowdsec.net/blog/multi-server-setup)) - there were 2 cases:
* after agent registration in _remote_ LAPI, local LAPI was disabled in config (as mentioned in article). After container restart it didn't start becase of the next errors
```
Error: no matches found
Generate local agent credentials
level=warning msg="crowdsec local API is disabled because 'enable' is set to false"
level=fatal msg="local API is disabled -- this command must be run on the local API machine"
```
* after agent registration in _remote_ LAPI, local LAPI was not disabled. After container restart LAPI URL was reverted to http://0.0.0.0:8080 according to the [entrypoint script](https://github.com/crowdsecurity/crowdsec/blob/master/docker/docker_start.sh#L251-L262)

Even if container started without any errors, this solution is not suitable for us - we use AWS Fargate as container runtime, so container restart would lead to its re-creation.

3. Start agent container with remote LAPI configured initially:
* token can't be passed in local_api_credentials.yaml
```
level=fatal msg="loading api client: failed unmarshaling api client credential configuration file '/etc/crowdsec/local_api_credentials.yaml': yaml: unmarshal errors:\n  line 1: field token not found in type csconfig.ApiCredentialsCfg"
```
* passing token via `password` parameter doesn't work (expected :smile:)

### Why is this needed?

It would be great to have possibility to pass autoregistration token to agent machines (either via local_api_credentials.yaml or via env vars)

**Comment by LaurenceJJones** [comment]
Please note the PR you are referencing is not general release yet, you can either use the `dev` container image OR shortly the `RC1` tag for `1.6.4`

I will read the rest of your issue, I thought I'll point this out early so you can see why it may not have worked.

**Comment by LaurenceJJones** [comment]
Just for documentation skake I dump some ideas here:

https://github.com/crowdsecurity/crowdsec/blob/5d414f58e5eda2659f44d17be9ea39da87f9e791/docker/docker_start.sh#L277-L284

We check if LAPI is disabled then set the configuration via env vars that are passed into the container, if we have a registration token these could be randomly generated instead 🤷🏻 so we should add an if check firstly if these vars exist then set them?

then there should be a if check on the registration key and if this is passed we run `cscli lapi register` either with the passed username/password or randomly generated? so for example I would expect the agent env vars to be below:

```
DISABLE_LOCAL_API=true
LOCAL_API_URL=http://remote_api_url
LOCAL_API_TOKEN=XYZ123
```

**Comment by LaurenceJJones** [comment]
Also because I never asked, are you using some orchestration software like kubernetes for auto scaling?

**Comment by ViktorOrda** [comment]
We use AWS ECS Fargate as orchestrator

**Comment by ViktorOrda** [comment]
IMO env for LAPI (i.e. `LOCAL_API_TOKEN`) is not so necessary since LAPI should be single - token can be passed via `config.yaml.local`. But for agents passing parameters as envs is more convenient

**Comment by LaurenceJJones** [comment]
> IMO env for LAPI (i.e. `LOCAL_API_TOKEN`) is not so necessary since LAPI should be single - token can be passed via `config.yaml.local`. But for agents passing parameters as envs is more convenient

Yeah I came to the same conclusion whilst I was editing my comments 👍🏻

**Comment by ViktorOrda** [comment]
> technically you could just modify the startup script
Was thinking about the same during investigation 🙂

**Comment by burylo** [comment]
I use Helm to set up CrowdSec. Agent's registration is ok, but validation didn't
wait-for-lapi-and-register container:
`level=info msg="Successfully registered to Local API (LAPI)"
level=info msg="Local API credentials written to '/etc/crowdsec/local_api_credentials.yaml'"
level=warning msg="Run 'sudo systemctl reload crowdsec' for the new configuration to be effective."`
crowdsec-agent container:
`time="2025-01-09T17:52:37Z" level=fatal msg="unable to start crowdsec routines: authenticate watcher (crowdsec-agent-9pgjh): API error: machine crowdsec-agent-9pgjh not validated"`

**Comment by fgeck** [comment]
@ViktorOrda I am running into the same issue as @burylo .
Should I open a new issue, or is there some documentation or similar because we are missing something obvious?

**Comment by burylo** [comment]
@fgeck Of course, why not? I searched the documentation, but I didn't find anything like that. However, something may have been updated during this time. I was looking for a solution for my project, and came across CrowdSec, it was almost perfect except for the problem with automatic agent validation. So I went ahead and rejected the CrowdSec option

**Comment by LaurenceJJones** [comment]
> [@ViktorOrda](https://github.com/ViktorOrda) I am running into the same issue as [@burylo](https://github.com/burylo) . Should I open a new issue, or is there some documentation or similar because we are missing something obvious?

Hey @fgeck and @burylo if you are using helm best place to raise issues is the chart repo itself https://github.com/crowdsecurity/helm-charts

The issue with agent auto registration is you need to define a token to be used by the LAPI to accept requests:

```
lapi:
  config.yaml.local: |
    api:
      server:
        auto_registration:
          enabled: true
          token: "${REGISTRATION_TOKEN}" # /!\ Do not modify this variable (auto-generated and handled by the chart)
          allowed_ranges:
            - "10.0.0.8/8"
```

**Comment by burylo** [comment]
No, problem with validation, bit registration. It's not very convenient to
go in and manually validate pods every time after they've been converted.

сб, 8 лют. 2025 р., 11:08 користувач Laurence Jones <
***@***.***> пише:

**Comment by LaurenceJJones** [comment]
> No, problem with validation, bit registration

If you define the token, it automatically validates.

**Comment by fgeck** [comment]
@LaurenceJJones thank you. Your post solved my issue. Flux was templating `"${REGISTRATION_TOKEN}"` to an empty string.
